<a href="https://colab.research.google.com/github/adrianprogrammer15/math/blob/main/Copia_de_practica_algebra.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Práctica 3. Álgebra
* Para aprender sobre Python vea la playlist [Python para Matemáticas](https://www.youtube.com/playlist?list=PL8gpDXwO-9l4HkDJ9QhB4H-cOiT0BYGau)
* Para obtener ideas y vídeos para la resolución de la práctica vea los archivos siguientes: [Ideas](https://github.com/financieras/math/tree/main/ideas)

## Práctica 3.1
**Determinante en Excel**  

* Cree un fichero en Excel de nombre algebra.xlsm.
* La extensión **xlsm** corresponde a los archivos de Excel habilitados para macros.
* Utilice la Hoja 1 para hacer la Práctica 3.1.
* Se pide:
 1. Cree una matriz cuadrada de dimensión 5×5, formada por números aleatorios enteros entre -9 y +9.
 2. Calcule el determinante de esta matriz
 3. Cree un programa en **VBA** que busque una matriz de este tipo cuyo determinante sea cero.
 4. Copie la matriz encontrada y péguela debajo con "pegado especial valores" para que no varíe su contenido al efectuar el recálculo manual de la hoja.

## Práctica 3.2
**Matriz inversa en Excel**  

* Use el mismo archivo de Excel con extensión **xlsm** del caso anterior. 
* Utilice la Hoja 2 para hacer la Práctica 3.2.
* Se pide:
 1. Crea la matriz $A$ que es una matriz cuadrada de dimensiones 5×5 usando números aleatorios enteros entre -9 y +9, excepto cero
 2. Calcule su determinante
 3. Calcule su transpuesta usando la función ```TRANSPONER```
 4. Calcule su matriz inversa $A^{-1}$
 5. Multiplique la matriz $A$ por su inversa $A^{-1}$ y compruebe que el producto es la matriz identidad.
 $$AA^{-1}=I$$

## Práctica 3.3
**Resolución de sistemas lineales en Excel**  

* Use el mismo archivo de Excel con extensión **xlsm** del caso anterior. 
* Utilice la Hoja 3 para hacer la Práctica 3.3.
* Se pide:
 1. Cree un sistema de 5 ecuaciones con 5 incógnitas con **solución única**.
 2. Una de las ecuaciones será: $\quad x_1+2x_2-x_3+x_4-3x_5=6$
 3. Muestre las ecuaciones del sistema usando el [editor de ecuaciones](https://support.microsoft.com/es-es/office/editor-de-ecuaciones-6eac7d71-3c74-437b-80d3-c7dea24fdf3f) de Excel. $\quad Ax=b$
 4. Monte la matriz $A$ del sistema de ecuaciones y la matriz $b$ con los términos independientes. Muestre ambas matrices.
 5. Calcule la matriz inversa $A^{-1}$
 6. Resuelva el sistema lineal de ecuaciones con este método: $\quad Ax=b \quad \Longrightarrow \quad x=A^{-1}b$
 7. Compruebe si el resultado obtenido es la solución del sistema de ecuaciones. Justifique esa comprobación.

## Práctica 3.4
**Matriz inversa en Python**  

Desarrolla un código en Python que al ejecutarse consiga los siguientes apartados.  
1. Crea una matriz cuadrada de dimensiones 3×3 usando números aleatorios enteros entre -9 y +9, excepto cero
2. Buscar un caso donde el determinante sea igual a 1
3. Si el determinante es 1 imprime la matriz y su determinante
4. Calcula e imprime su transpuesta
5. Calcula e imprime su matriz inversa
6. Multiplica la matriz creada por su inversa y comprueba que el producto es la matriz identidad.
7. Calcula cuantos segundos ha tardado en realizarse el proceso.

## Solución 3.4

In [ ]:
import numpy as np

n = 3   

while True:
    m = np.random.randint(-9,10, size=(n, n))   
    determ = np.linalg.det(m)
    if determ == 1:
        print("La matriz m es: ")
        print(m)
        break
transpuesta = np.transpose(m)
inversa = np.linalg.inv(m)
print(f"El determinante de la matriz es {determ}")
print("La matriz traspuesta de m es: "  )
print(transpuesta)
print("La inversa de la matriz m es: ")
print(inversa)

identidad = np.dot(m, np.linalg.inv(m))
print("La multiplicacion de la matriz inversa y la matriz m es: ")
print(identidad)   

from time import process_time

t1 = process_time()

for i in range(10_000_000):
    cuadrado = i**2

t2 = process_time()
print(f"El proceso ha tardado {t2-t1} segundos.")






La matriz m es: 
[[ 8  1 -6]
 [ 7  1  9]
 [-7 -1 -8]]
El determinante de la matriz es 1.0
La matriz traspuesta de m es: 
[[ 8  7 -7]
 [ 1  1 -1]
 [-6  9 -8]]
La inversa de la matriz m es: 
[[   1.   14.   15.]
 [  -7. -106. -114.]
 [   0.    1.    1.]]
La multiplicacion de la matriz inversa y la matriz m es: 
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
El proceso ha tardado 3.4464222929999977 segundos.


## Práctica 3.5
**Resolución de sistemas lineales en Python**  

1. Cree un sistema de 4 ecuaciones con 4 incógnitas con **solución única**.
2. La solución del sistema es: $\; \begin{pmatrix}x_1, x_2, x_3, x_4\end{pmatrix} = \begin{pmatrix} 3, 4, 5, 2\end{pmatrix}$  
3. Una de las ecuaciones del sistema es: $\enspace x_1+3x_2-x_3-2x_4=6$  
4. El vector $b$ de términos independientes es: $\enspace b=\begin{pmatrix}6, 3, 5, 7\end{pmatrix}$  
5. Usando números aleatorios elija los coeficientes que faltan de la matriz $A$ del sistema de ecuaciones.
  - Los coeficientes deben ser números enteros entre -5 y +5, excluido el cero.
  - Imprima la matriz $A$.  
6. Calcule e imprima el determinante de $A$.
7. Calcule e imprima la matriz inversa $A^{-1}$  
8. Resuelva el sistema lineal de ecuaciones con este método: $$Ax=b \quad \Longrightarrow \quad x=A^{-1}b$$
9. Compruebe que el resultado obtenido es la solución del sistema de ecuaciones.  
10. Resuelva nuevamente el sistema de ecuaciones usando ```np.linalg.solve```.

## Solución 3.5

Comenzamos generando listas de 4 números aleatorios (-5 al 5, menos el 0) para construir nuestro sistema.

In [ ]:
import numpy as np
list = np.random.choice([-5, -4, -3, -2, -1, 1, 2, 3, 4, 5], 4)
print(list)

[ 1  4  2 -5]


Una vez elegidos los grupos de 4 números para conformar nuestro sistemas de ecuaciones, construimos una matriz 4x4 con estos valores. Imprimimos la matriz A:

In [ ]:
fila1 = [1, 3, -1,-2]
fila2 = [-5,-1 , 4 , 4]

fila3 = [ 1, -4 ,-1 , 4]

fila4 = [ 1,  4,  2, -5]
matriz = [f1, f2, f3, f4]

lista = np.array(matriz)
print(lista)

[[ 1  3 -1 -2]
 [-5 -1  4  4]
 [ 1 -4 -1  4]
 [ 1  4  2 -5]]


Calculamos el determinante de $A$

In [ ]:
import numpy as np
m = [[ 1, 3, -1, -2], [-5, -1, 4, 4], [ 1, -4, -1, 4] , [ 1, 4, 2, -5] ]
np.linalg.det(m)
int(np.linalg.det(m))     # Convertimos el número en entero debido a que nos arroja gran cantidad de decimales residuales

-111

Calculamos  $A^{-1}$  

In [ ]:
import numpy as np

m = np.array([[ 1, 3, -1, -2], [-5, -1, 4, 4], [ 1, -4, -1, 4] , [ 1, 4, 2, -5] ])
print("Matriz:")
print(m)
print()

print("Inversa:")
minv = np.linalg.inv(m)
print(minv)




Matriz:
[[ 1  3 -1 -2]
 [-5 -1  4  4]
 [ 1 -4 -1  4]
 [ 1  4  2 -5]]

Inversa:
[[ 0.26126126  0.02702703  0.53153153  0.34234234]
 [ 0.56756757  0.16216216  0.18918919  0.05405405]
 [-0.02702703  0.13513514  0.32432432  0.37837838]
 [ 0.4954955   0.18918919  0.38738739  0.06306306]]


Buscamos la solución del sistema mediante la ecuación  $$Ax=b \quad \Longrightarrow \quad x=A^{-1}b$$

In [ ]:
import numpy as np
a = np.array([[ 1, 3, -1, -2], [-5, -1, 4, 4], [ 1, -4, -1, 4] , [ 1, 4, 2, -5] ])
b = np.array([6,3,5,7])
a_inv = np.linalg.inv(a)
np.dot(a_inv, b) 
round_off_values = print(np.round_(np.dot(a_inv, b)) ) # Para redondear con numpy trabajamos con round_off_values e imprimimos la solución del sistema con números enteros



[7. 5. 5. 6.]


En este caso resolveremos por el método ```np.linalg.solve```.

In [ ]:
import numpy as np
a = np.array([[ 1, 3, -1, -2], [-5, -1, 4, 4], [ 1, -4, -1, 4] , [ 1, 4, 2, -5] ])
b = np.array([6,3,5,7])
solucion = np.linalg.solve(a,b)
round_off_values = print(np.round_(solucion))    # Volvemos a redondear

[7. 5. 5. 6.]


## Práctica 3.6
**Resolución de sistemas lineales con Octave**  

1. Tome el mismo sistema 4x4 con solución única del ejercicio anterior.
2. Muestra el sistema usando $\LaTeX$
3. Resuélvelo con Octave por dos métodos
4. ¿Ambos métodos dan el mismo resultado?

## Solución 3.6

$$\left.
\begin{align*}
x + 3y- z -2t &=  6 \\
-5x -y +4z +4t&=  3 \\
x -4y -1z +4t &= 5 \\
x +4y +2z -5t &= 7
\end{align*}
\right\}$$

In [ ]:
!apt install octave

In [ ]:
!apt update

##Método 1

In [ ]:
%%writefile primero.m
A = [1 3 -1 -2; -5 -1 4 4; 1 -4 -1 4; 1 4 2 -5]
b = [6; 3; 5; 7]
x= inv(A)*b
% solución 1, rats expresa los número como fracciones o enteros
s1 =rats(x)

Writing primero.m


In [ ]:
!octave primero.m

octave: X11 DISPLAY environment variable not set
octave: disabling GUI features
A =

   1   3  -1  -2
  -5  -1   4   4
   1  -4  -1   4
   1   4   2  -5

b =

   6
   3
   5
   7

x =

   6.7027
   5.2162
   4.5135
   5.9189

s1 =

     248/37
     193/37
     167/37
     219/37



## Método 2

In [ ]:
%%writefile tercero.m
A = [1 3 -1 -2; -5 -1 4 4; 1 -4 -1 4; 1 4 2 -5]
b = [6; 3; 5; 7]
x = A\b
% solución 3 
s3 = rats(x)

Writing tercero.m


In [ ]:
!octave tercero.m

octave: X11 DISPLAY environment variable not set
octave: disabling GUI features
A =

   1   3  -1  -2
  -5  -1   4   4
   1  -4  -1   4
   1   4   2  -5

b =

   6
   3
   5
   7

x =

   6.7027
   5.2162
   4.5135
   5.9189

s3 =

     248/37
     193/37
     167/37
     219/37



- La solución redondeada tanto en el primer caso como en el segundo caso es $ (7 , 5 , 5 , 6)$. 
Y efectivamente las soluciones sí coinciden en ambos casos

## Práctica 3.7
**Diagonalización en Python**

* Sea la matriz cuadrada $A$ de orden 4x4

$$A=\left(
\begin{array}{rrrr}
 -1 &-2 &  3 &-1 \\
  0 & 2 &  0 & 0 \\
  0 & 0 &  2 & 0 \\
  0 & 0 &  0 & 1
\end{array}
\right)$$
1. Calcula los **autovalores** y **autovectores**
2. Calcula y muestra:
 - la matriz diagonal $D$
 - la matriz de paso $P$
 - su inversa $P^{-1}$  
3. Comprueba que se cumple que $A^3=PD^3P^{-1}$

## Solución 3.7

Con la librería ````Numpy```` obtenemos las matrices P y D y los autovalores y autovectores

In [ ]:
import numpy as np
A = np.array([[-1,-2,3,-1], [0,2,0,0], [0,0,2,0], [0,0,0,1]])
d,P = np.linalg.eig(A)
print("d: ", d)
D = np.diag(d)
print("\nD:\n", D)
print("\nP:\n", P)

d:  [-1.  2.  2.  1.]

D:
 [[-1.  0.  0.  0.]
 [ 0.  2.  0.  0.]
 [ 0.  0.  2.  0.]
 [ 0.  0.  0.  1.]]

P:
 [[ 1.         -0.5547002   0.70710678 -0.4472136 ]
 [ 0.          0.83205029  0.          0.        ]
 [ 0.          0.          0.70710678  0.        ]
 [ 0.          0.          0.          0.89442719]]


## Utilizamos la librería ````Sympy```` que muestra más clara la matriz P y D 

In [ ]:
from sympy import Matrix
A = Matrix([[-1,-2,3,-1], [0,2,0,0], [0,0,2,0], [0,0,0,1]])
P, D = A.diagonalize()
D

Matrix([
[-1, 0, 0, 0],
[ 0, 1, 0, 0],
[ 0, 0, 2, 0],
[ 0, 0, 0, 2]])

In [ ]:
P

Matrix([
[1, -1, -2, 1],
[0,  0,  3, 0],
[0,  0,  0, 1],
[0,  2,  0, 0]])

## Calculamos $A^3$

In [ ]:
import numpy as np
A = np.array([[-1,-2,3,-1], [0,2,0,0], [0,0,2,0], [0,0,0,1]])
print(A@A@A)

[[-1 -6  9 -1]
 [ 0  8  0  0]
 [ 0  0  8  0]
 [ 0  0  0  1]]


## Y calculamos $PD^3P^{-1}$ 

In [ ]:
import numpy as np
A = np.array([[-1,-2,3,-1], [0,2,0,0], [0,0,2,0], [0,0,0,1]])
d,P = np.linalg.eig(A)
print("d: ", d)
D = np.diag(d)
print("\nD:\n", D)
print("\nP:\n", P)
Dcubo = D @ D @ D
print("\nDcubo:\n", Dcubo)
print("\nSolución:\n", P @ Dcubo @ np.linalg.inv(P))

d:  [-1.  2.  2.  1.]

D:
 [[-1.  0.  0.  0.]
 [ 0.  2.  0.  0.]
 [ 0.  0.  2.  0.]
 [ 0.  0.  0.  1.]]

P:
 [[ 1.         -0.5547002   0.70710678 -0.4472136 ]
 [ 0.          0.83205029  0.          0.        ]
 [ 0.          0.          0.70710678  0.        ]
 [ 0.          0.          0.          0.89442719]]

Dcubo:
 [[-1.  0.  0.  0.]
 [ 0.  8.  0.  0.]
 [ 0.  0.  8.  0.]
 [ 0.  0.  0.  1.]]

Solución:
 [[-1. -6.  9. -1.]
 [ 0.  8.  0.  0.]
 [ 0.  0.  8.  0.]
 [ 0.  0.  0.  1.]]


- ### Como podemos observar, la igualdad se cumple.

## Práctica 3.8
**Diagonalización con Octave**

Resuelva el ejercicio anterior, pero en esta ocasión utilizando Octave.

## Solución 3.8

In [ ]:
%%writefile diagonaliza.m
A = [-1,-2,3,-1; 0,2,0,0; 0,0,2,0; 0,0,0,1]
[P,D]=eig(A)
Dcubo = D^3
Acubo = A^3
PxDcuboXPinv = P*Dcubo*P^-1

Overwriting diagonaliza.m


In [ ]:
!octave diagonaliza.m


octave: X11 DISPLAY environment variable not set
octave: disabling GUI features
A =

  -1  -2   3  -1
   0   2   0   0
   0   0   2   0
   0   0   0   1

P =

   1.00000  -0.55470   0.70711  -0.44721
   0.00000   0.83205   0.00000   0.00000
   0.00000   0.00000   0.70711   0.00000
   0.00000   0.00000   0.00000   0.89443

D =

Diagonal Matrix

  -1   0   0   0
   0   2   0   0
   0   0   2   0
   0   0   0   1

Dcubo =

Diagonal Matrix

  -1   0   0   0
   0   8   0   0
   0   0   8   0
   0   0   0   1

Acubo =

  -1  -6   9  -1
   0   8   0   0
   0   0   8   0
   0   0   0   1

PxDcuboXPinv =

  -1  -6   9  -1
   0   8   0   0
   0   0   8   0
   0   0   0   1



## Práctica 3.9
**Libertad para elegir el lugar de trabajo**  
Resolver con **Octave**.

Una empresa de 1.000 empleados permite que sus trabajadores cada día acudan a trabajar a la oficina o que realicen su trabajo desde su casa. Inicialmente el 90% de la plantilla trabajaba en la oficina y solo el 10% lo hacían en casa. Se ha observado que los que un día trabajan en casa tienen una probabilidad del 70% de realizarlo también desde casa al día siguiente, pero que en el restante 30% de los casos acuden a la oficina al día siguiente. Por el contrario, se ha observado que los que un día trabajaron en la oficina al día siguiente vuelven
a la oficina en el 60% de los casos y en el 40% restante van a su casa al día siguiente.
* Determinar qué número de empleados se quedan en casa transcurridos dos días.
* ¿Y a largo plazo cuantos se quedan en casa?

## Solución 3.9

In [ ]:
%%writefile proceso.m
empleados = 1000
% state vector
inicial = [.9; .1]
% transition probability matrix
A = [.6, .3; .4, .7]
[P,D]=eig(A)
Dcuadrado = D^2
Acuadrado = A^2
PxDcuadradoXPinv = P*Dcuadrado*P^-1
dia2 = Acuadrado * inicial
disp("Nº de empleados que se quedan en casa el segundo día: "), disp(dia2(2)*empleados)
D10 = D^10
A10 = P*D10*P^-1
dia10 = A10 * inicial
disp("Nº de empleados que se quedan en casa a largo plazo son: "), disp(dia10(2)*empleados)

Writing proceso.m


In [ ]:
!octave proceso.m


octave: X11 DISPLAY environment variable not set
octave: disabling GUI features
empleados =  1000
inicial =

   0.90000
   0.10000

A =

   0.60000   0.30000
   0.40000   0.70000

P =

  -0.70711  -0.60000
   0.70711  -0.80000

D =

Diagonal Matrix

   0.30000         0
         0   1.00000

Dcuadrado =

Diagonal Matrix

   0.090000          0
          0   1.000000

Acuadrado =

   0.48000   0.39000
   0.52000   0.61000

PxDcuadradoXPinv =

   0.48000   0.39000
   0.52000   0.61000

dia2 =

   0.47100
   0.52900

Nº de empleados que se quedan en casa el segundo día: 
 529
D10 =

Diagonal Matrix

   5.9049e-06            0
            0   1.0000e+00

A10 =

   0.42857   0.42857
   0.57143   0.57143

dia10 =

   0.42857
   0.57143

Nº de empleados que se quedan en casa a largo plazo son: 
 571.43


### - Al transcurrir dos días, el número de trabajadores que se quedan en casa será de 529.
### - La tendencia de trabajadores que deciden trabajar en casa es de 571.43 trabajadores

## Práctica 3.10
**Tendencia del mercado**  
Resolver con **Python**.

En el mercado de automóviles de gran lujo únicamente se disputan la clientela dos marcas (A y B). Las ventas son $M=100.000$ automóviles anuales, no variando esta cifra con el transcurso de los años. La marca A es líder de mercado en este momento con el 75% de las ventas. Sus directivos desean consolidar su liderazgo e incluso aumentarlo, para lo cual todos los años efectúan una fuerte campaña publicitaria con la que consiguen cada año un 40% de los que fueron clientes de su competidora, aunque la empresa A pierde un 20% de sus clientes que se van a la competencia.
* ¿Cuál es la situación a los 3 años?
* ¿Y al cabo de mucho tiempo?

## Solución 3.10

In [ ]:
import numpy as np

X0 = np.array([.75, .25])
A = np.array([[.8, .4], [.2 ,.6]])
X3 = A@A@A@X0
print("Situación a los tres años:", X3)

d,P = np.linalg.eig(A)
D = np.diag(d)
Pinv = np.linalg.inv(P)
A10 = P@(D**10)@Pinv
X10 = A10@X0
print("Situación a los diez años:", X10)
print("Situación a los veinte años:", P@(D**20)@Pinv@X0)

Situación a los tres años: [0.672 0.328]
Situación a los diez años: [0.6666754 0.3333246]
Situación a los veinte años: [0.66666667 0.33333333]


### - La situación del mercado al cabo de **3 años** es la siguiente: 67,2% una empresa y 32,8% la otra empresa.
### - La situación a los **10 años** :Una empresa obtiene el 66,67% del mercado y su competencia 33,33%
### - Finalmente al cabo de **20 años**, la clientela se ha estabilizado manteniendo valores iguales a los de la situación de hace 10 años.